### ISSUE TITLE
Difference between query_roc_auc and sklearn's roc_auc_score

### ISSUE THREAD

**Issue by matthiasdold** [abstractqqq/polars_ds_extension#376]
With the following minimal example, I get a difference (0.3 vs 0.47) for calculating the `roc_auc` value, while the true/false-positive rates agree... Since `fpr` and `tpr` agree, I would assume a potential bug/inconsitency in the calculation of the trapezoidal integration. 

I can reconstruct the `polars_ds` value when I ignore the first segment of the trapezoid (skipping index=0 in the sum below).

-> Unfortunately, I do not know Rust, but I would assume that the index range could be changed somewhere [here](https://github.com/abstractqqq/polars_ds_extension/blob/49839af10a5bed0dce305926c75d66797cde1136/src/num_ext/trapz.rs#L7), and the point at x=0 and y=0 would need to be added.

-----------------------------------
Minimal example:

```python
import polars as pl
import polars_ds as pds
from scipy.integrate import trapezoid
from sklearn.metrics import roc_auc_score, roc_curve

data = {
    "ytrue": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
    "ypred": [1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1],
}
dt = pl.DataFrame(data).with_columns(
    pl.lit("test").alias("aux")
)

dtg = dt.group_by("aux").agg(
    pds.query_roc_auc("ytrue", "ypred").alias("roc_auc"),
    pds.query_tpr_fpr("ytrue", "ypred").alias("tpr_fpr"),
)

# For the simple binary case (with labels {0, 1}), `roc_auc_score` reduces to this:
# https://github.com/scikit-learn/scikit-learn/blob/da08f3d99194565caaa2b6757a3816eef258cd70/sklearn/metrics/_ranking.py#L43
fpr, tpr, th = roc_curve(dt["ytrue"], dt["ypred"])
trapezoid(tpr, fpr)

# or in one go from sklearn.metrics
roc_auc_score(dt["ytrue"], dt["ypred"])

# replicating the current `polars_ds` implementatation by skipping index=0
def manual_trapz(y, x):
    """Most simple implementation, but skipping the index at 0"""
    return sum((y[i] + y[i + 1]) * (x[i + 1] - x[i]) / 2 for i in range(1, len(y) - 1))
```

The output on my machine:
```python
In [13]: print(dtg)
shape: (1, 3)
┌──────┬──────────┬─────────────────────────────────┐
│ aux  ┆ roc_auc  ┆ tpr_fpr                         │
│ ---  ┆ ---      ┆ ---                             │
│ str  ┆ f64      ┆ list[struct[3]]                 │
╞══════╪══════════╪═════════════════════════════════╡
│ test ┆ 0.309091 ┆ [{0,1.0,1.0}, {1,0.545455,0.6}… │
└──────┴──────────┴─────────────────────────────────┘

In [14]: print(f"{fpr=}, {tpr=}, {th=}")
fpr=array([0. , 0.6, 1. ]), tpr=array([0.        , 0.54545455, 1.        ]), th=array([inf,  1.,  0.])

In [15]: trapezoid(tpr, fpr)
Out[15]: np.float64(0.47272727272727266)

In [16]: roc_auc_score(dt["ytrue"], dt["ypred"])
Out[16]: np.float64(0.47272727272727266)

In [17]: manual_trapz(tpr, fpr)
Out[17]: np.float64(0.3090909090909091)

```

------------------------
MacOS, M1, Python 3.13, with the following libraries:
```
polars                       1.27.1
polars_ds                 0.10.0
scipy                         1.15.2
scikit-learn                1.6.1
```

**Comment by abstractqqq** [comment]
You are actually almost there..

The difference is that in sklearn, the following has a third value (th = inf, with tpr=0 and fpr=0)

`fpr, tpr, th = roc_curve(dt["ytrue"], dt["ypred"])`

which serves as a padding on the left when you integrate using trapz. If we add this zero padding to pds's implementation:
```
from scipy.integrate import trapezoid

data = {
    "ytrue": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
    "ypred": [1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1],
}
dt = pl.DataFrame(data).with_columns(
    pl.lit("test").alias("aux")
)

temp = dt.select(
    pds.query_tpr_fpr("ytrue", "ypred").alias("tpr_fpr"),
).unnest("tpr_fpr")

- trapezoid(list(temp["tpr"]) + [0.0], list(temp["fpr"]) + [0.0])

```
will return 0.47272727272727266.


This is something that I honestly think doesn't matter in practice - it is really just up to the user because trapezoidal integration is just an approximation anyways. But I guess I lack the authority to make any such claims..

Well, I will make an update. Thank you for pointing this out!

**Comment by abstractqqq** [comment]
First a small correction, the range index in your manual trapz is off by 1. The correct version should be: 
```
def manual_trapz(y, x):
    """Most simple implementation, but skipping the index at 0"""
    return sum((y[i + 1] + y[i]) * (x[i + 1] - x[i]) / 2 for i in range(0, len(y) - 1))
```

You are actually 99% close to the truth of the matter. If you examine the output of `fpr, tpr, th = roc_curve(dt["ytrue"], dt["ypred"])`, you see that it has three values per vector, and the one that causes the difference is when th = inf, fpr = 0 and tpr = 0. This value serves as some padding on the left, but pds's tpr_fpr doesn't return this arbitrary value.

I want to say trapezoidal integration is just an approximation anyways.. But I guess it also makes sense to pad with this 0.0. 

Thank you for pointing this out!

**Comment by abstractqqq** [comment]
> Thanks a lot [@abstractqqq](https://github.com/abstractqqq) for addressing this! The `manual_trapz` was indexed this way on purpose, to replicate the `polars_ds` behavior.
> 
> And yes the point at (0, 0), equivalent to a threshold of `inf`, would need to be added by the calculations in rust. Although it is just an approximation, I think this point is very relevant to include, especially if there are only a limited number of thresholds

Thanks again. It is a good catch and thanks for providing all the details so that I can fix this so quickly. But please bear with me for a moment and let me clarify why the `manual_trapz` should start with `range(0, ..)` instead of `range(1, ..)`

The Rust code in question is the following:
```rust
cfavml::add_vector(&y[1..], &y[..y.len() - 1], &mut y_d);
```

If you translate it to NumPy:
```python
y[1:] + y[:-1]
```
You see the first sum in the vec really is y[1] + y[0], and so it needs to start with 0.
